### ISSUE TITLE
Loss of precision of error location

### ISSUE THREAD

**Issue by CohenCyril** [coq/coq#12992]
#### Description of the problem

Between Coq 8.9 and now, there has been a considerable loss in precision in locating the source of an error in some scripts.
e.g. the following script fails on purpose for coq >= 8.9, but with worse and worse error message:

```coq
Lemma test m n : S m = n -> m = 4.
Proof. 
intro e; rewrite e.
```
- In Coq 8.7.x, 8.8.x, and 8.9.1:
  ```
  Toplevel input, characters 9-18:
  > intro e; rewrite e.
  >          ^^^^^^^^^
  Error: Found no subterm matching "S m" in the current goal.
  ```
  This is not as precise as it could be (the culprit is `e`), but at least it is not a regression.

- Coq 8.10.2:
  ```
  Toplevel input, characters 0-19:
  > intro e; rewrite e.
  > ^^^^^^^^^^^^^^^^^^^
  Error: Found no subterm matching "S m" in the current goal.
  ```
  This is a regression, the error is less precise than before.

- Coq 8.11.2 and 8.12.0:
  ```
  Toplevel input, characters 0-7:
  > intro e; rewrite e.
  > ^^^^^^^
  Error: Found no subterm matching "S m" in the current goal.
  ```
  Now this is a bug, the error message in the end is right, but the cuplrit is not `intro e` so the location is wrong.
  This looks loosely related to #12837.

#### Coq Version
Coq >= 8.10

**Comment by CohenCyril** [comment]
FYI @gares @maximedenes @ybertot

**Comment by gares** [comment]
Even if the root cause could be the same (a choice point wrongly set), this has nothing to do with SSR, right?

**Comment by CohenCyril** [comment]
> Even if the root cause could be the same (a choice point wrongly set), this has nothing to do with SSR, right?

Yes, the problem occurs without loading anything (no more than the prelude)

**Comment by ppedrot** [comment]
Wasn't that fixed by https://github.com/coq/coq/pull/12774?

**Comment by CohenCyril** [comment]
Maybe... I did not check.

**Comment by SkySkimmer** [comment]
So is this fixed now?

**Comment by ybertot** [comment]
As far as I am concerned, the problem is fixed in the master branch of Coq.  I hope the fix will be included in coq-8.12.1

**Comment by ybertot** [comment]
Well, I just checked.  The fix is indeed tagged for coq-8.12.1.
